### PR TITLE
Refactor Genome Browser track panel and drawer

### PR DESCRIPTION
**Pull request by azangru**
## Related JIRA Issue(s)
- https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1389
- https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1404

### Change redux slice responsible for genome browser drawer
- Change the shape of the redux slice responsible for the genome browser drawer, such that the name of the drawer view is co-located with the data required to render this view
- Refactor the drawer redux slice to use redux-toolkit

### Refactor TrackPanelList items
Instead of a single `TrackPanelItem` component, create purely presentational components to maintain the layout, as well as higher-order components for different types of tracks (`TrackPanelGene`, `TrackPanelTranscript`, `TrackPanelTranscriptGroup`, and `TrackPanelRegularItem` (see the [diagram](https://github.com/Ensembl/ensembl-client/files/7476702/Track.panel.elements.pdf)).

### New functionality
- When track panel gene is collapsed, show the number of transcripts that this gene has.

## TODO
- re-examine `EnsObjectTrack` type; how much of it can be removed?
- consider removing `isDrawerOpen` from drawer state
- switch `GeneSummary` and `TranscriptSummary` from Apollo client to redux-toolkit query
- move drawer reducer into the genome browser reducer

## Deployment URL
http://refactor-gb-drawer.review.ensembl.org

## Views affected
Genome Browser page

### CODE REVIEW

File: src/ensembl/src/content/app/browser/state/drawer/types.ts [hunk 1]
```diff
@@ -14,27 +14,33 @@
...
+    geneId: string; // in ensObjectId format
+  };
+};
 
-export type DrawerState = Readonly<{
-  [genomeId: string]: DrawerStateForGenome;
-}>;
+export type TranscriptDrawerView = {
+  name: 'transcript_summary';
+  payload: {
```
Comment (msmanoj) [inline src/ensembl/src/content/app/browser/state/drawer/types.ts:30]: Why do we need `payload` instead of having IDs at the same level as name?

**Comment by jyothishnt** [comment]
Please could you confirm if this is the correct behaviour? This is the behaviour in internal-2020 as well

1. Open track panel drawer
2. click on first transcript summary link and open transcript download
3. Now click on the next transcript
You will see download opened. 

4. Now click on gene summary and you will see download closed
5. Now click on transcripts and you will see download closed

File: src/ensembl/src/content/app/browser/track-panel/track-panel-list/track-panel-items/TrackPanelGene.tsx [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        isCollapsed={isCollapsed}
+        visibilityStatus={trackStatus}
+        onChangeVisibility={() =>
+          toggleTrack({ trackId: GENE_TRACK_ID, status: trackStatus })
+        }
+        onShowMore={onShowMore}
+        toggleExpand={toggleExpand}
+      >
+        <div className={styles.label}>
+          <span className={styles.labelTextStrong}>
```
Comment (msmanoj) [inline src/ensembl/src/content/app/browser/track-panel/track-panel-list/track-panel-items/TrackPanelGene.tsx:155]: `labelTextStrong` -> `labelStrong`?

Comment (azangru) [reply src/ensembl/src/content/app/browser/track-panel/track-panel-list/track-panel-items/TrackPanelGene.tsx:155]: In the styles, there is also `.labelTextSecondary` and `.labelTextSecondaryStrong`. Without "text", `labelSecondary` doesn't sound quite right to me.

File: src/ensembl/src/content/app/browser/state/drawer/types.ts (hunk 1 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/state/drawer/types.ts:30]: I was imitating redux actions.

**Comment by azangru** [comment]
> Please could you confirm if this is the correct behaviour? This is the behaviour in internal-2020 as well \
> Open track panel drawer \
> click on first transcript summary link and open transcript download \
>Now click on the next transcript \
>You will see download opened. \
> Now click on gene summary and you will see download closed \
> Now click on transcripts and you will see download closed

I don't know if this is correct, but it sounds like something out of scope of this PR.

File: src/ensembl/src/content/app/browser/state/drawer/drawerSlice.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    state[genomeId] = { ...defaultDrawerStateForGenome };
+  }
+};
+
+const drawerSlice = createSlice({
+  name: 'genome-browser-drawer',
+  initialState: {} as DrawerState,
+  reducers: {
+    // TODO: discuss whether we need this part of the state at all;
+    // isn't { drawerView: null } sufficient to represent a closed drawer?
```
Comment (msmanoj) [inline src/ensembl/src/content/app/browser/state/drawer/drawerSlice.ts:109]: I think `drawerView: null` is sufficient.

File: src/ensembl/src/content/app/browser/state/drawer/types.ts (hunk 1 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/state/drawer/types.ts]: Agreed to move the contents of the payload field to the top level; done.

File: src/ensembl/src/content/app/browser/state/drawer/drawerSlice.ts (hunk 3 above)
Comment (azangru) [reply src/ensembl/src/content/app/browser/state/drawer/drawerSlice.ts]: ✅

File: src/ensembl/src/content/app/browser/browserSelectors.ts [hunk 4]
```diff
@@ -52,6 +54,35 @@ export const getBrowserActiveEnsObject = (state: RootState) => {
...
+        tracksGroup: 'commonTracks';
+      }
+  )
+) => {
+  const { genomeId, tracksGroup, categoryName, trackId } = params;
+  const allBrowserTrackStates = getBrowserTrackStates(state);
+  const savedTrackStatus =
+    tracksGroup === 'objectTracks'
+      ? allBrowserTrackStates?.[genomeId]?.[tracksGroup]?.[
+          (params as any).objectId
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/browser/browserSelectors.ts]: ✅

File: src/ensembl/src/content/app/browser/state/drawer/drawerSelectors.ts [hunk 5]
```diff
@@ -15,8 +15,8 @@
  */
 
 import { RootState } from 'src/store';
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/browser/state/drawer/drawerSelectors.ts]: ✅

**Comment by ens-ma7** [comment]
@imransl and @msmanoj its ready for another review.

**Review by msmanoj — APPROVED** [review]

**Pull request merged**